[PGO] Add option to always instrumenting loop entries

ronryvchin opened the pull request:
This patch extends the PGO infrastructure with an option to prefer the instrumentation of loop entry blocks.
This option is a generalization of https://github.com/llvm/llvm-project/commit/19fb5b467bb97f95eace1f3637d2d1041cebd3ce,
and helps to cover cases where the loop exit is never executed.
An example where this can occur are event handling loops.

Note that change does NOT change the default behavior.

Dinistro commented on llvm/include/llvm/Transforms/Instrumentation/CFGMST.h:
```diff
@@ -154,6 +159,9 @@ template <class Edge, class BBInfo> class CFGMST {
           }
           if (BPI != nullptr)
             Weight = BPI->getEdgeProbability(&BB, TargetBB).scale(scaleFactor);
+          if (InstrumentLoopEntries && LI != nullptr &&
```
Nit: Maybe explain in a small comment what is happening here.

Dinistro reviewed (commented):
Dropped a bunch of nit comments. The general idea makes sense to me and I think we can soonish open this for general reviews of people experienced with these components.

Dinistro reviewed (commented):
Nice work 🙂 
Dropped one more nit comment. Apart from this, this should be ready for a review of owners/maintainers of the PGO infra.

ellishg commented on llvm/lib/Transforms/Instrumentation/PGOInstrumentation.cpp:
```diff
@@ -1980,8 +1995,11 @@ PreservedAnalyses PGOInstrumentationGen::run(Module &M,
   };
+  auto LookupLI = [&FAM](Function &F) {
+    return &FAM.getResult<LoopAnalysis>(F);
+  };
```
Honestly I'm not sure why we don't simply pass the `FAM` rather than these lambdas, but this probably shouldn't block your work.

ellishg commented on llvm/include/llvm/Transforms/Instrumentation/CFGMST.h:
```diff
@@ -291,10 +301,14 @@ template <class Edge, class BBInfo> class CFGMST {
+           "expected a LoopInfo to instrumenting loop entries");
     buildEdges();
     sortEdgesByWeight();
     computeMinimumSpanningTree();
```
I wonder if it makes sense to add an assert here that each loop entry will be instrumented.

Dinistro replied:
Could you elaborate what exactly you mean by this? Assert in the end that all loop entries are instrumented or what exactly?

ellishg replied:
If `InstrumentLoopEntries` is true, then each loop entry should be instrumented. I'm asking if we should add a compile time assert that indeed each loop entry will be instrumented after we compute the minimum spanning tree.

ellishg commented on llvm/test/Transforms/PGOProfile/loop3.ll:
```diff
@@ -0,0 +1,62 @@
+; RUN: opt %s -passes=pgo-instr-gen -pgo-instrument-loop-entries=false -S | FileCheck %s --check-prefixes=GEN,NOTLOOPENTRIES
```
Can you also add a test that we can consume these profiles with loop entries?

ellishg reviewed (commented):
I think we also need to add a new variant mask for instrumenting loop entries. Otherwise we won't be able to match the profiles they generate correctly

https://github.com/llvm/llvm-project/blob/dd78d7c7be5b8948cf5841e8033e59adebf230ad/llvm/lib/Transforms/Instrumentation/PGOInstrumentation.cpp#L439-L441

Dinistro commented on llvm/include/llvm/ProfileData/InstrProfReader.h:
```diff
@@ -123,6 +123,8 @@ class InstrProfReader {
 
   virtual bool instrEntryBBEnabled() const = 0;
 
+  virtual bool instrLoopEntriesEnabled() const = 0;
```
I think it would be sensible to add a comment to this function, even when it's somewhat trivial.

ellishg commented on compiler-rt/include/profile/InstrProfData.inc:
```diff
@@ -742,9 +744,10 @@ serializeValueProfDataFrom(ValueProfRecordClosure *Closure,
-#define VARIANT_MASK_IR_PROF (0x1ULL << 56)
-#define VARIANT_MASK_CSIR_PROF (0x1ULL << 57)
-#define VARIANT_MASK_INSTR_ENTRY (0x1ULL << 58)
+#define VARIANT_MASK_IR_PROF (0x1ULL << 55)
+#define VARIANT_MASK_CSIR_PROF (0x1ULL << 56)
+#define VARIANT_MASK_INSTR_ENTRY (0x1ULL << 57)
```
To maintain backwards compatibility these bits cannot be changed. I guess we need to expand the mask to support more bits.

ronryvchin replied:
Are bits 33-55 in use?

ronryvchin replied:
I see that you are right (historically it started at `56` and climbed higher), but I cannot find usages of bit `55`

Dinistro replied:
Considering that there is no usage of bit `55` and that there is no documentation about this being reserved, I would argue that we should be free to use `55`. If any downstream user is relying on such an undocumented implementation detail, we should be free to break it.

ellishg commented on llvm/test/Transforms/PGOProfile/Inputs/thinlto_indirect_call_promotion.profraw:
This seems to be unused.

ellishg commented on llvm/test/Transforms/PGOProfile/loop3.ll:
```diff
@@ -0,0 +1,78 @@
+; RUN: opt %s -passes=pgo-instr-gen -pgo-instrument-loop-entries=false -S | FileCheck %s --check-prefixes=GEN,NOTLOOPENTRIES
+; RUN: llvm-profdata merge %S/Inputs/loop3.proftext -o %t.profdata
```
Instead of creating new files in the `Inputs` directory, lets use `split-file` to inline the `.proftext` files into this test.

https://llvm.org/docs/TestingGuide.html#extra-files

ellishg replied:
Also, it might be cleaner to split this into two separate tests: `PGOProfile/loop-entry-gen.ll` and `PGOProfile/loop-entry-use.ll`.

ellishg commented on llvm/test/Transforms/PGOProfile/loop3.ll:
```diff
@@ -0,0 +1,78 @@
+for.entry:
+; GEN: for.entry:
+; LOOPENTRIES: call void @llvm.instrprof.increment(ptr @__profn_test_simple_for_with_bypass, i64 {{[0-9]+}}, i32 3, i32 1)
+; NOTLOOPENTRIES-NOT: call void @llvm.instrprof.increment
```
You can add the `--implict-check-not="@llvm.instrprof.increment"` to your `FileCheck` command to reduce the number of check lines needed.

ellishg commented:
> @ellishg Do you know how `llvm-project/llvm//test/tools/llvm-profdata/Inputs/function-entry-coverage.profdata` was generated? The new bit broke `llvm-project/llvm/test/tools/llvm-profdata/show-covered.test` so I need to re-generate its input, but I cannot find the generation logic. It was added in [11d3074](https://github.com/llvm/llvm-project/commit/11d3074267324e811ade248a991564fa6722754e)

Thanks for asking. I think you should not need to change this test, especially since this flag is off by default. I suspect it is broken because the variant mask bits were changed.

ronryvchin committed:
[PGO] Supporting code for always instrumenting loop entries

This patch extends the PGO infrastructure with an option to prefer the instrumentation of loop entry blocks.
This option is a generalization of https://github.com/llvm/llvm-project/commit/19fb5b467bb97f95eace1f3637d2d1041cebd3ce,
and helps to cover cases where the loop exit is never executed.
An example where this can occur are event handling loops.

Note that change does NOT change the default behavior.

ellishg commented on llvm/test/Transforms/PGOProfile/loop_entries_gen.ll:
```diff
@@ -0,0 +1,58 @@
+; RUN: opt %s -passes=pgo-instr-gen -pgo-instrument-loop-entries=false -S | FileCheck %s --check-prefixes=CHECK,NOTLOOPENTRIES --implicit-check-not=@llvm.instrprof.increment
+; RUN: opt %s -passes=pgo-instr-gen -pgo-instrument-loop-entries=true -S | FileCheck %s --check-prefixes=CHECK,LOOPENTRIES --implicit-check-not=@llvm.instrprof.increment
+; RUN: opt %s -passes=pgo-instr-gen -pgo-instrument-entry=true -S | FileCheck %s --check-prefixes=CHECK,FUNCTIONENTRY --implicit-check-not=@llvm.instrprof.increment
```
```suggestion
; RUN: opt %s -passes=pgo-instr-gen -S | FileCheck %s --check-prefixes=CHECK,NOTLOOPENTRIES --implicit-check-not=@llvm.instrprof.increment
; RUN: opt %s -passes=pgo-instr-gen -pgo-instrument-loop-entries -S | FileCheck %s --check-prefixes=CHECK,LOOPENTRIES --implicit-check-not=@llvm.instrprof.increment
; RUN: opt %s -passes=pgo-instr-gen -pgo-instrument-entry -S | FileCheck %s --check-prefixes=CHECK,FUNCTIONENTRY --implicit-check-not=@llvm.instrprof.increment
```

ellishg commented on llvm/test/Transforms/PGOProfile/loop_entries_use.ll:
```diff
@@ -0,0 +1,106 @@
+; RUN: rm -rf %t && split-file %s %t
+
+; RUN: llvm-profdata merge %t/default.proftext -o %t/default.profdata
+; RUN: opt %t/main.ll -passes=pgo-instr-use -pgo-instrument-loop-entries=false -pgo-test-profile-file=%t/default.profdata -S | FileCheck %s
+; RUN: llvm-profdata merge %t/loop_entries.proftext -o %t/loop_entries.profdata
+; RUN: opt %t/main.ll -passes=pgo-instr-use -pgo-instrument-loop-entries=true -pgo-test-profile-file=%t/loop_entries.profdata -S | FileCheck %s
+; RUN: llvm-profdata merge %t/function_entry.proftext -o %t/function_entry.profdata
+; RUN: opt %t/main.ll -passes=pgo-instr-use -pgo-instrument-entry=true -pgo-test-profile-file=%t/function_entry.profdata -S | FileCheck %s
```
You do not need to specify the flags when consuming profile because that info in encoded in the header.

```suggestion
; RUN: opt %t/main.ll -passes=pgo-instr-use -pgo-test-profile-file=%t/default.profdata -S | FileCheck %s
; RUN: llvm-profdata merge %t/loop_entries.proftext -o %t/loop_entries.profdata
; RUN: opt %t/main.ll -passes=pgo-instr-use -pgo-test-profile-file=%t/loop_entries.profdata -S | FileCheck %s
; RUN: llvm-profdata merge %t/function_entry.proftext -o %t/function_entry.profdata
; RUN: opt %t/main.ll -passes=pgo-instr-use -pgo-test-profile-file=%t/function_entry.profdata -S | FileCheck %s
```

ronryvchin commented:
@ellishg Thank you for your comments!
I think everything is addressed now, is there anything else that should be done before merging?

ellishg reviewed (approved):
LGTM
